In [ ]:
from bertopic import BERTopic
import pandas as pd
import numpy as np
import time

# 1. Topic Modelling using BERTopic

In [3]:
df = pd.read_csv("../data/cleaned/cleaned_reviews.csv")
embeddings = np.load("../data/cleaned/sentence_embeddings.npy")

In [34]:
df_copy = df.copy()

In [6]:
len(df)

30157

In [7]:
embeddings.shape[0]

30157

In [4]:
start = time.time()
topic_model = BERTopic(min_topic_size=50,language='english')
topics,probs = topic_model.fit_transform(df['text_clean_full'].tolist(),embeddings=embeddings)  
end = time.time()

In [5]:
print(f"Number of topics: {len(topic_model.get_topic_info())}")
print(f"Time taken: {(end-start):.2f} seconds")

Number of topics: 77
Time taken: 40.83 seconds


In [6]:
topic_info = topic_model.get_topic_info()
print(topic_info.head(20))

    Topic  Count                                   Name  \
0      -1  11173                  -1_the_case_screen_it   
1       0   3540           0_charge_charger_charging_it   
2       1   1400                 1_watch_band_wrist_the   
3       2   1094                     2_phone_to_and_sim   
4       3    750  3_bubbles_protector_screen_protectors   
5       4    545      4_fingerprint_reader_finger_print   
6       5    529             5_cards_card_wallet_credit   
7       6    510           6_broke_broken_after_product   
8       7    465           7_buttons_button_volume_case   
9       8    417                8_stylus_pen_tip_pencil   
10      9    394          9_working_stopped_worked_work   
11     10    358             10_stick_stay_adhesive_off   
12     11    332                 11_fit_phone_didnt_did   
13     12    321                   12_color_case_the_is   
14     13    309       13_case_protection_protect_phone   
15     14    308              14_case_broke_apart_cases 

In [7]:
start = time.time()
topic_model_v2 = BERTopic(min_topic_size=30,language='english')
topics_v2,probs_v2 = topic_model_v2.fit_transform(df['text_clean_full'].tolist(),embeddings=embeddings)
end = time.time()

In [10]:
topic_info_v2 = topic_model_v2.get_topic_info()

In [12]:
print(f"Number of topics: {len(topic_model_v2.get_topic_info())} ")
print(f"Time taken: {(end-start):.2f} seconds")
print(f"Noise cluster size: {topic_info_v2[topic_info_v2['Topic']==-1]['Count'].values[0]}")

Number of topics: 127 
Time taken: 7.20 seconds
Noise cluster size: 11773


In [13]:
topic_info_v2 = topic_model_v2.get_topic_info()
print(topic_info_v2.head(20))

    Topic  Count                                    Name  \
0      -1  11773                   -1_the_screen_case_it   
1       0   2695         0_charge_charger_charging_cable   
2       1    977                      1_phone_sim_to_you   
3       2    556       2_fingerprint_reader_finger_print   
4       3    523              3_cards_card_wallet_credit   
5       4    494            4_broke_broken_after_product   
6       5    453            5_buttons_button_volume_side   
7       6    416                 6_stylus_pen_tip_pencil   
8       7    413           7_working_stopped_worked_work   
9       8    350                8_case_broke_apart_cases   
10      9    339   9_protector_cracked_screen_protectors   
11     10    322       10_magnet_magnetic_strong_magnets   
12     11    306              11_sound_ear_headset_noise   
13     12    300         12_protection_protect_case_drop   
14     13    285              13_watch_face_apple_screen   
15     14    276                 14_colo

In [14]:
topic_model.reduce_topics(df['text_clean_full'].tolist(),nr_topics=40)
reduced_info = topic_model.get_topic_info()

In [15]:
print(f"Number of topics after reduction: {len(reduced_info)}")
print(reduced_info.head(20))

Number of topics after reduction: 40
    Topic  Count                                    Name  \
0      -1  11173                      -1_the_it_and_case   
1       0   4774                      0_charge_to_it_and   
2       1   1573          1_screen_protector_the_bubbles   
3       2   1527                   2_the_cards_it_magnet   
4       3   1520                        3_case_the_is_it   
5       4   1478                     4_watch_band_the_it   
6       5   1086                    5_fit_iphone_not_for   
7       6    864                6_broke_cracked_it_after   
8       7    791                      7_stick_dash_to_it   
9       8    551                 8_screen_cracked_my_the   
10      9    544           9_working_stopped_work_worked   
11     10    465           10_buttons_button_case_volume   
12     11    417                   11_stylus_pen_tip_the   
13     12    401                   12_sound_hear_ear_the   
14     13    321                   13_fit_size_wrong_too   
15 

In [32]:
print(reduced_info)

    Topic  Count                                    Name  \
0      -1  11834                        -1_the_it_and_to   
1       0   2961            0_charge_charger_charging_it   
2       1   1601                       1_the_cards_it_to   
3       2   1471                     2_watch_band_the_it   
4       3   1311                        3_case_the_is_it   
5       4   1203                      4_phone_to_and_the   
6       5   1140                  5_fit_iphone_not_phone   
7       6    891                   6_stick_mount_dash_to   
8       7    854      7_fingerprint_screen_protector_the   
9       8    596           8_screen_cracked_my_protector   
10      9    559             9_broke_cracked_apart_after   
11     10    546          10_working_stopped_work_worked   
12     11    542                    11_color_case_the_is   
13     12    500                   12_sound_hear_ear_the   
14     13    481           13_buttons_button_case_volume   
15     14    420                   14_st

# 2. Topic mapping

In [16]:
df = df.drop(columns=['topic_id','topic_label'],errors='ignore')

In [18]:
topic_labels = {
    -1: "Noise",
    0: "Charging issues",
    1: "Ring/Card Slot Cases",
    2: "Wearable bands",
    3: "Phone cases",
    4: "Phone issue",
    5: "Case Fitting issues", #Redundant with topic 16
    6: "Mounts,Holder issue",
    7: "Fingerprints/screen protector",
    8: "Cracked screen/Display",
    9: "Broken item", # Redundant with topic 10
    10: "Item stopped working",
    11: "Case Color related",
    12: "Sound/Hearing issues",
    13: "Issue with buttons/Buttons not working",
    14: "Stylus/Pen issues",
    15: "Screen protector issue (bubble formation)",
    16: "Fitting issues", 
    17: "Camera/Camera lens issues",
    18: "Vent mounts/holder issues",
    19: "Item does not work", # Redundant with topic 10
    20: "Return related issue",
    21: "Scratch issues",
    22: "Popsockets",
    23: "Build Quality/Form Factor complaints",
    24: "Product Quality issues",
    25: "Glass Screen related issues",
    26: "Irrelevant",
    27: "Phone Covers/Protection",
    28: "Irrelevant",
    29: "Irrelevant",
    30: "Irrelevant",
    31: "Tripod issues",
    32: "Plastic product",
    33: "Not a genuine product",
    34: "Crystals lost",
    35: "Not reliable",
    36: "Irrelevant",
    37: "Installation issues",
    38: "Device Fitting issues" #Redundant with topic 16
}

In [23]:
df['topic_id'] = topic_model.topics_
df['topic_label'] = df['topic_id'].map(topic_labels)
print(df[['text','topic_id','topic_label']].head(10))
print(df['topic_label'].value_counts().head(15))

                                                text  topic_id  \
0  Putting it on a night stand drawer or top of h...         7   
1  A bit complicated to apply. Requires spraying ...         4   
2  This product is cheaply made, some of the part...         6   
3  I absolutely do not like this phone. Screen is...         0   
4  Refurbished so I expected it to work well but ...         0   
5                   WRONG CABLE. . .DID NOT USE. . .         0   
6  I didn’t like this product because water got i...         3   
7  Got it in the mail and was way too excited to ...         0   
8  Fell apart right away.  Clamps get in the way ...        10   
9  Within 2 weeks of using it fell off the phone ...         2   

                     topic_label  
0  Fingerprints/screen protector  
1                    Phone issue  
2            Mounts,Holder issue  
3                Charging issues  
4                Charging issues  
5                Charging issues  
6                    Phone c

In [24]:
print(df['topic_label'].isna().sum()) 
print(df['topic_id'].nunique()) 

0
40


In [25]:
df.to_csv("../data/cleaned/cleaned_reviews.csv", index=False)

In [37]:
charging_complaints = df[df['topic_label'] == 'Charging issues'][['rating', 'text']].head(5)
print(charging_complaints)

    rating                                               text
5      2.0                   WRONG CABLE. . .DID NOT USE. . .
7      1.0  Got it in the mail and was way too excited to ...
12     2.0  I had one that lasted for about a year. I boug...
20     2.0  The USB port just broke loose the 3rx time I u...
36     1.0  UPDATE:  purchased  6/20/18  .....  it only ch...


## Topic Modeling — BERTopic

### Configuration & Tuning

| Version | min_topic_size | Topics | Noise Count | Noise % |
|---------|---------------|--------|-------------|---------|
| v1 | 50 | 79 | 11,173 | 37% |
| v2 | 30 | 127 | 11,773 | 33% |
| Final (reduced) | 50 → reduced | 40 | 11,834 | 39% |

**Decision:** v1 with `reduce_topics(nr_topics=40)` selected — fewer, cleaner topics
with slightly lower noise vs v2. 40 topics is actionable for a support routing system.

**Why pre-computed embeddings?**
Passed `sentence_embeddings.npy` directly to BERTopic — skips redundant embedding
step, saves ~200 seconds per run. BERTopic ran UMAP + HDBSCAN only (58 seconds).

### Topic Quality
- **34 actionable topics** covering: Charging Issues, Wearable Bands, Screen Protector
  Bubbles, Fingerprint Reader, Stylus Issues, Car Vent Mounts, Camera Accessories, etc.
- **5 irrelevant clusters** identified and labeled: Spanish reviews (topic 28),
  meta-commentary (topic 26), generic bad reviews (topic 29, 30), bad cluster (topic 36)
- **3 redundant topic pairs** noted as merge candidates for production:
  - Topics 5, 16, 38 → all fitting-related
  - Topics 9, 10, 19 → all broken/stopped working

### Topic Distribution (Top 5)
| Topic | Count |
|-------|-------|
| Noise | 11,834 |
| Charging Issues | 2,746 |
| Ring/Card Slot Cases | 1,392 |
| Wearable Bands | 1,062 |
| Irrelevant | 877 |

### Known Limitations
- **39% noise** — genuinely ambiguous short reviews ("Not working!", "One Star")
  that lack distinctive vocabulary for clustering. These remain available for
  semantic retrieval but are excluded from topic-based routing.
- Some legitimate complaints fell into irrelevant clusters (e.g. row 2:
  *"product is cheaply made, parts missing"* → topic 36/Irrelevant)
- Non-English reviews (Spanish) formed their own cluster — not filtered in preprocessing

### Next Step
Topic labels stored as `topic_id` and `topic_label` columns in `cleaned_review.csv`
— will be used as metadata in ChromaDB (Phase 5) for filtered retrieval.

## Topic Labelling Fix

**Root cause**: `reduce_topics(nr_topics=40)` was called on topic_model, but the dataframe was populated using `topic_model.topics_` — which reflects the pre-reduction 79-topic assignment. The reduced topic IDs only live on the model object after `reduce_topics()` is called; they are not automatically written back. As a result, `cleaned_reviews.csv` was saved with 78 unique topic IDs instead of 40, and ChromaDB was ingested with those stale labels.

**Detected when**: topic_centroids computed in Phase 6 returned 78 centroids instead of the expected 40.

**Fix**: Re-ran the label assignment using the post-reduction topics, verified `df['topic_id'].nunique() == 40` and `df['topic_label'].isna().sum() == 0`, re-saved CSV, deleted and re-ingested ChromaDB collection.

**Lesson**: After calling `reduce_topics()`, always re-extract topics via `topic_model.topics_` in the same cell, since the attribute is updated in-place. Verify with `nunique()` before saving.